### PR TITLE
Shadow DOM with regular Element fallback -- requires custom Mocha rep…

### LINKED ISSUE #452: Tests don't show on Chrome; show on Firefox but don't show results
#### Issue Description

The menu button for testing doesn't show on Chrome, but shows on Firefox. On Firefox, tests work, but when I click to view details about passes/fails, there is no text in the pop-up window.

#### Browser Information

* Browser Name, Version: Firefox Quantum 61.0.1. // Google Chrome 67.0.3396.99
* Operating System: Windows 10
* Mobile, Desktop, or Tablet: Desktop

#### Your Code / Link to Your Pen

```
 https://codepen.io/dtujmer/pen/QxPBdz
 
 
```

#### Screenshot

image

### PR DESCRIPTION
**Pull request by tbushman**
#### Pre-Submission Checklist

- [x] Your pull request targets the `master` branch of freeCodeCamp/testable-projects-fcc
- [x] Branch starts with either `fix/`, `feature/`, or `translate/` (e.g. `fix/signin-issue`)
- [x] You have only one commit (if not, [squash](http://forum.freecodecamp.com/t/how-to-squash-multiple-commits-into-one-with-git/13231) them into one commit).
- [x] Your changes have been tested either locally or using a newly created CDN based on your fork's testable-projects-fcc/build/bundle.js file
- [x] Working changes have been added to the build by running `npm run build`

#### Type of Change

- [x] Small bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds new functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)
- [ ] Add new translation (feature adding new translations)

#### Checklist:

- [x] Tested changes locally.

- [x] Closes currently open issue: Closes #452 #470 #473 #476 #501

#### Description

In order to use Shadow DOM, we need a custom reporter so that we can programatically append test results to the Mocha modal (Mocha can't find the `#mocha` element inside the Shadow DOM. When Shadow DOM is not supported in a browser, a regular `<div>` is used. See notes in code.

### CODE REVIEW

**Comment by tbushman** [comment]
I need to adapt the Selenium tests now, but if you want to try running this locally, I would appreciate it. Thank you.

**Comment by tbushman** [comment]
See https://github.com/SeleniumHQ/selenium/issues/5869

**Comment by tbushman** [comment]
Update: was able to query using `webdriver.By.js()` and automated tests are working.

**Comment by tchaffee** [comment]
@tbushman I ran this locally and all the tests pass. I'm travelling at the moment so I'm not sure when I'll get to the code review. But I'm excited about this change - it's a much better approach to isolate our CSS from Camper project CSS.  What did you do differently to get the shadow DOM to work this time around?

**Comment by tchaffee** [comment]
@tbushman I would actually get rid of the CSS overrides. We don't support Firefox so as long as it works in Chrome for now we are good. I'm not sure when I'll get to the code review, but it's on the top of my list - just a bit difficult to squeeze in with my travel right now.

File: src/index.js [hunk 1]
```diff
@@ -52,18 +50,44 @@ import createTreeMapTests from './project-tests/tree-map-tests';
...
 
 let projectNameLocal = false;
+// filter out unnecessary stack trace in Mocha reporter
+const filterStack = stackTraceFilter();
+// Wrapper for shadow DOM
+const testDiv = document.createElement('div');
+testDiv.setAttribute('id', 'fcc_test_suite_wrapper');
+document.body.appendChild(testDiv);
+// Using a shadow DOM, the fCC css won't interfere with student css
+// A fallback div is provided.
```
Comment (tchaffee) [inline src/index.js]: Can we get rid of the fallback div? We are only supporting recent versions of Chrome for now.

File: src/index.js [hunk 2]
```diff
@@ -272,14 +323,86 @@ export function FCCResetTests(suite) {
...
+    create a new section with testList */
+  if (!parentTitle || parentTitle !== test.parent.title) {
+    count++;
+    parentTitle = test.parent.title;
+    var testSection = document.createElement('li');
+    testSection.setAttribute('class', `fcc_section_${count} suite`);
+    var parentTitleNode = document.createElement('h1');
+    parentTitleNode.innerText = parentTitle;
+    testSection.appendChild(parentTitleNode);
+    /* Currently using <ul>, but if we switch to <ol> we can eliminate all of
```
Comment (tchaffee) [inline src/index.js]: This is a good idea. I will create an issue for this. Can you just change this comment to start `/* TODO: Currently using...`

File: src/stylesheets/mocha-modal.css [hunk 3]
```diff
@@ -100,7 +100,7 @@ div#mocha{
   -o-border-radius:5px;
   border-radius:5px;
 }
-#mocha .test.pass.fast .duration{
+#mocha .test.pass.fast .duration, #mocha .test.fail .duration{
```
[the remark this answers is not in the document]
Comment (tchaffee) [reply src/stylesheets/mocha-modal.css]: Yes please!

**Comment by tbushman** [comment]
@tchaffee Have you looked at the reporter running? Let me know what you think of it now. I'll do the CSS changes in a separate PR. This CSS was the minimum required for a custom reporter.  
It's safer to keep the fallback div, and it works.

**Comment by tchaffee** [comment]
@tbushman 

> I'll do the CSS changes in a separate PR. This CSS was the minimum required for a custom reporter.

Agreed, better in a separate PR.

> It's safer to keep the fallback div, and it works.

Ok, it's your code and your decision and the code works. Just as an FYI, I was going by one of my favorite coding principals: https://en.wikipedia.org/wiki/You_aren't_gonna_need_it with that suggestion. Might be interesting for you to experiment with going forward.

> Have you looked at the reporter running? Let me know what you think of it now. 

I did. It looks great.

**Comment by tchaffee** [comment]
@tbushman One more question before merge. The console gets the output of the test result. That wasn't happening in the previous version. Is there a way to suppress this? Or maybe a better question: do we want to keep it? The only risk is that it being a change and different behavior there is a small risk of it breaking something for some Campers. I can't think of what that case would be. Just past experience telling me "don't fix it if it ain't broken" :-D

selection_189

**Comment by tbushman** [comment]
@tchaffee OK, changing the reporter type to `base` omits the reporter text from the console. Thanks!  
Also, I'm erring on the side of letting the test suite remain working in Firefox. You've mentioned YAGNI before. Thanks.

**Comment by tchaffee** [comment]
>  changing the reporter type to base omits the reporter text from the console. 

Excellent. LGTM! Merging.

**Comment by tchaffee** [comment]
Hmmm, that doesn't sound right. "Localhost only" would be a weird restriction. Let's investigate and see what we can find.

**Comment by tchaffee** [comment]
I just looked at a CodePen and it looks like it supports ShadowDOM to me. If you have time, we can hop on Skype or hangouts and talk it through so I can better understand what you are looking at.

**Comment by tchaffee** [comment]
@tbushman I don't know where to start with automatically creating the bundle either. But you just reminded me that should be a top priority because it will save us time and also get the bundle out of the repo so we don't have to deal with merge problems anymore. I'll look into automating it, unless you really wanted to work on that?

**Pull request merged**